src/WebUI.jl

```diff
@@ -0,0 +1,508 @@
+        pr === nothing && return html(500, "Registration failed: Making pull request failed")
+
+        url = web_url(pr)
+        html("""Registry PR successfully created, see it <a href="$url" target="_blank">here</a>!""")
+    else
+        html(500, "Registration failed: " * branch.error)
+    end
+end
+
+HTTP.@register ROUTER "GET" ROUTE_INDEX error_handler(index)
+HTTP.@register ROUTER "GET" ROUTE_AUTH error_handler(auth)
+HTTP.@register ROUTER "GET" ROUTE_CALLBACK error_handler(callback)
+HTTP.@register ROUTER "GET" ROUTE_SELECT error_handler(select)
+HTTP.@register ROUTER "POST" ROUTE_REGISTER error_handler(register)
+
+##############
+# Entrypoint #
+##############
+
+function init_providers()
+    disabled = split(get(ENV, "DISABLED_PROVIDERS", ""))
+
+    if !in("github", disabled)
+        PROVIDERS["github"] = Provider(;
+            name="GitHub",
+            client=GitHubAPI(;
+                url=get(ENV, "GITHUB_API_URL", GitHub.DEFAULT_URL),
+                token=Token(ENV["GITHUB_API_TOKEN"]),
+            ),
+            client_id=ENV["GITHUB_CLIENT_ID"],
+            client_secret=ENV["GITHUB_CLIENT_SECRET"],
+            auth_url=get(ENV, "GITHUB_AUTH_URL", "https://github.com/login/oauth/authorize"),
+            token_url=get(ENV, "GITHUB_TOKEN_URL", "https://github.com/login/oauth/access_token"),
+            scope="public_repo",
+        )
+    end
+
+    if !in("gitlab", disabled)
+        PROVIDERS["gitlab"] = Provider(;
+            name="GitLab",
+            client=GitLabAPI(;
+                url=get(ENV, "GITLAB_API_URL", GitLab.DEFAULT_URL),
+                token=PersonalAccessToken(ENV["GITLAB_API_TOKEN"]),
+            ),
+            client_id=ENV["GITLAB_CLIENT_ID"],
+            client_secret=ENV["GITLAB_CLIENT_SECRET"],
+            auth_url=get(ENV, "GITLAB_AUTH_URL", "https://gitlab.com/oauth/authorize"),
+            token_url=get(ENV, "GITLAB_TOKEN_URL", "https://gitlab.com/oauth/token"),
+            scope="read_user",
+            include_state=false,
+            token_type=OAuth2Token,
+        )
+    end
+
+    haskey(ENV, "EXTRA_PROVIDERS") && include(ENV["EXTRA_PROVIDERS"])
+end
+
+function init_registry()
+    url = ENV["REGISTRY_URL"]
+    k = get(ENV, "REGISTRY_PROVIDER") do
+        if occursin("github", url)
+            "github"
+        elseif occursin("gitlab", url)
+            "gitlab"
+        end
+    end
+    haskey(PROVIDERS, k) || error("Unsupported registry host")
+    forge = PROVIDERS[k].client
+    owner, name = splitrepo(url)
+    repo = @gf get_repo(forge, owner, name)
+    repo === nothing && error("Registry lookup failed")
+    REGISTRY[] = Registry(forge, repo, url, get(ENV, "REGISTRY_CLONE_URL", url))
+end
+
+# server should be created with Sockets.listen if you want to be able to close it.
+# Most of the time you should only set either of these options in testing.
+function main(; init::Bool=true, server=nothing)
```

Comment:
I don't like environment variables, since they are a bit of a hidden state. IP and PORT should be arguments, either keyword, or positional with default. The `main` function should be callable from the shell, and thus should not take any Julia types as arguments -- only strings and numbers. So `server` should not be an argument to main. Extract bits out into maybe a `startServer` method if you need to have `server` as an argument.

Reply:
> I don't like environment variables

Let's agree to disagree :sweat_smile:.

If we don't want environment variables, then we need a config file I guess.

Reply:
FWIW, I've found that running deployed software with environment variable configuration + systemd or [supervisord](http://supervisord.org/) to be very effective

Reply:
this needs to be started by a shell script. so it will be started as `julia -e "using Registrator; WebUI.main()"` -- So we need to pass the port from the shell script to the julia function. So I was saying make it a parameter of main, then then we can call it as 

`julia -e "using Registrator; WebUI.main(;port=$PORT)"` from a shell script. 

No config file needed

Reply:
That's not the only configuration though. GitHub/GitLab application information/API keys, registry URL, etc. A quick `grep` for `ENV` or browse through the setup docs should yield them all.

Reply:
Ah OK, indeed. I take back some of that back -- the tokens will need some type of a secret manager, and so environment variables are best.  So keep them as is. 

I'd still recommend changing ip/port. they need to be handled by the start scripts, and doing it like I suggest above will be more flexible, IMO.

Reply:
Moved `ip` and `port` to `main` keyword arguments.